pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go

```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
+		},
+		"overlapping spans with close ts without quantization": {
+			withFrontierQuantization: false,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("d")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+			},
+			expectedFrontierEntries: 3,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(12)},
+		},
+		"non-overlapping spans with close ts with quantization": {
+			withFrontierQuantization: true,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("z")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12),
+			},
+			expectedFrontierEntries: 4,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+		"non-overlapping spans with close ts without quantization": {
+			withFrontierQuantization: false,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("z")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12),
+			},
+			expectedFrontierEntries: 6,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			frontier, err := span.MakeFrontier(tc.spans...)
+			require.NoError(t, err)
+			const quantize = time.Duration(10)
+			for _, e := range tc.events {
+				if tc.withFrontierQuantization {
+					e.Checkpoint.ResolvedTS = mockQuantization(e.Checkpoint.ResolvedTS, quantize)
+				}
+				_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
+				require.NoError(t, err)
+			}
+			frontier.Entries(func(sp roachpb.Span, ts hlc.Timestamp) span.OpResult {
+				t.Logf("span: %v, ts: %v\n", sp, ts)
+				return false
+			})
+			require.Equal(t, tc.expectedFrontierEntries, frontier.Len())
+			require.Equal(t, tc.expectedFrontier, frontier.Frontier())
+		})
+	}
+}
+
+func TestFrontierQuantizationRand(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	mockQuantization := func(ts hlc.Timestamp, quantize time.Duration) hlc.Timestamp {
+		ts.Logical = 0
+		ts.WallTime -= ts.WallTime % int64(quantize)
+		return ts
+	}
+
+	makeCheckpointEvent := func(key []byte, endKey []byte, ts int, logical int32) *kvpb.RangeFeedEvent {
+		return &kvpb.RangeFeedEvent{
+			Checkpoint: &kvpb.RangeFeedCheckpoint{
+				Span:       roachpb.Span{Key: key, EndKey: endKey},
+				ResolvedTS: hlc.Timestamp{WallTime: int64(ts), Logical: logical},
+			},
+		}
+	}
```

Comment:
nit these can be pulled out between these two tests

Reply:
Done.